File: ruby/lib/rerout/qr_options.rb
# frozen_string_literal: true

module Rerout
  # Options for the QR endpoint. All fields are optional.
  #
  # - `size` (Integer): module size in px (1..32). Server default 8.
  # - `margin` (Integer): quiet zone in modules (0..16). Server default 4.
  # - `ecc` (String): error correction level. One of `L`, `M`, `Q`, `H`.
  # - `domain` (String): force the QR to encode a specific verified custom domain.
  # - `refresh` (String, true): cache-bust on regenerate. `true` is serialized as `"1"`.
  class QrOptions
    attr_reader :size, :margin, :ecc, :domain, :refresh

    ALLOWED_ECC = %w[L M Q H].freeze

    def initialize(size: nil, margin: nil, ecc: nil, domain: nil, refresh: nil)
      if ecc && !ALLOWED_ECC.include?(ecc.to_s)
        raise ArgumentError, "ecc must be one of #{ALLOWED_ECC.inspect}, got #{ecc.inspect}"
      end

      @size = size
      @margin = margin
      @ecc = ecc
      @domain = domain
      @refresh = refresh
      freeze
    end

    # @return [Boolean] true when no field is set.
    def empty?
      size.nil? && margin.nil? && ecc.nil? && domain.nil? && refresh.nil?
    end

    # Serialize options as a `{ key => string }` hash ready to be turned into
    # a query string. `refresh: true` becomes `"1"`.
    def to_query_hash
      out = {}
      out['size'] = size.to_s unless size.nil?
      out['margin'] = margin.to_s unless margin.nil?
      out['ecc'] = ecc.to_s unless ecc.nil?
      out['domain'] = domain.to_s unless domain.nil?
      unless refresh.nil?
        out['refresh'] = refresh == true ? '1' : refresh.to_s
      end
      out
    end
  end
end

File: dotnet/src/Rerout/Models/QrOptions.cs
using System.Collections.Generic;

namespace Rerout.Models;

/// <summary>
/// QR rendering parameters for <see cref="Resources.Qr.Url"/> and
/// <see cref="Resources.Qr.SvgAsync"/>.
/// </summary>
/// <remarks>
/// All fields are optional — omit any to use the server default.
/// </remarks>
public sealed record QrOptions
{
    /// <summary>Module size in pixels. 1–32. Server default: 8.</summary>
    public int? Size { get; init; }

    /// <summary>Quiet-zone modules. 0–16. Server default: 4.</summary>
    public int? Margin { get; init; }

    /// <summary>Error-correction level — <c>L</c>, <c>M</c>, <c>Q</c>, or <c>H</c>.</summary>
    public string? Ecc { get; init; }

    /// <summary>
    /// Force the QR to encode a specific verified custom domain instead of the
    /// project's default branding domain.
    /// </summary>
    public string? Domain { get; init; }

    /// <summary>
    /// Cache-bust token. Setting <see cref="RefreshAlways"/> sends
    /// <c>refresh=1</c>; any other non-empty string is forwarded verbatim and
    /// triggers a fresh render server-side.
    /// </summary>
    public string? Refresh { get; init; }

    /// <summary>
    /// Whether to force a fresh render. When <c>true</c>, emits <c>refresh=1</c>
    /// unless an explicit <see cref="Refresh"/> token is also supplied (the
    /// token wins).
    /// </summary>
    public bool RefreshAlways { get; init; }

    /// <summary>Render this options bag into URL query pairs.</summary>
    internal IReadOnlyList<KeyValuePair<string, string>> ToQueryParameters()
    {
        var pairs = new List<KeyValuePair<string, string>>();
        if (Size is { } size)
        {
            pairs.Add(new KeyValuePair<string, string>("size", size.ToString(System.Globalization.CultureInfo.InvariantCulture)));
        }

        if (Margin is { } margin)
        {
            pairs.Add(new KeyValuePair<string, string>("margin", margin.ToString(System.Globalization.CultureInfo.InvariantCulture)));
        }

        if (!string.IsNullOrEmpty(Ecc))
        {
            pairs.Add(new KeyValuePair<string, string>("ecc", Ecc));
        }

        if (!string.IsNullOrEmpty(Domain))
        {
            pairs.Add(new KeyValuePair<string, string>("domain", Domain));
        }

        if (!string.IsNullOrEmpty(Refresh))
        {
            pairs.Add(new KeyValuePair<string, string>("refresh", Refresh));
        }
        else if (RefreshAlways)
        {
            pairs.Add(new KeyValuePair<string, string>("refresh", "1"));
        }

        return pairs;
    }
}
